### PR TITLE
ranger: choose more prefix columns when building ranges

### PR DESCRIPTION
**Pull request by eurekaka**
### What problem does this PR solve? 

Fix https://github.com/pingcap/tidb/issues/10027

### What is changed and how it works?

Return `eqOrInCount` instead of `eqCount` in `detachCNFCondAndBuildRangeForIndex` for subsequent detach.

### Check List 

Tests 

 - Unit test

Code changes

N/A

Side effects

N/A

Related changes

N/A

### CODE REVIEW

File: util/ranger/detacher.go [hunk 1]
```diff
@@ -48,7 +48,6 @@ func detachColumnCNFConditions(sctx sessionctx.Context, conditions []expression.
 		accessConditions = append(accessConditions, cond)
 		if checker.shouldReserve {
 			filterConditions = append(filterConditions, cond)
-			checker.shouldReserve = checker.length != types.UnspecifiedLength
```
Comment (lamxTyler) [inline util/ranger/detacher.go]: Why remove it?

File: util/ranger/detacher.go [hunk 2]
```diff
@@ -194,7 +185,7 @@ func detachCNFCondAndBuildRangeForIndex(sctx sessionctx.Context, conditions []ex
 	res.Ranges = ranges
 	res.AccessConds = accessConds
 	res.RemainedConds = filterConds
-	res.EqCondCount = eqCount
+	res.EqCondCount = eqOrInCount
```
Comment (lamxTyler) [inline util/ranger/detacher.go]: Could we rename the `EqCondCount`?

Comment (winoros) [inline util/ranger/detacher.go]: `EqCondCount` is used when checking the physical property pushed down to data source.
We cannot let it also contain the count of `IN` function.

Comment (qw4990) [reply util/ranger/detacher.go]: `EqCondCount` is used in `getIndexCandidate` to check if a index path match a property and in `splitCorColAccessCondFromFilters` to extract cor-columns to push down.
It seems safe in these two place if we make `EqCondCount` contain `IN` expressions.

**Pull request merged**